parsl/providers/azure/azure.py

```diff
@@ -0,0 +1,478 @@
+    from azure.mgmt.compute.models import DiskCreateOption
+    from msrestazure.azure_exceptions import CloudError
+
+    _api_enabled = True
+
+except ImportError:
+    _api_enabled = False
+else:
+    _api_enabled = True
+
+translate_table = {
+    'VM pending': 'PENDING',
+    'VM running': 'RUNNING',
+    'VM deallocated': 'COMPLETED',
+    'VM stopping': 'COMPLETED',  # We shouldn't really see this state
+    'VM stopped': 'COMPLETED',  # We shouldn't really see this state
+}
+
+
+class AzureProvider(ExecutionProvider, RepresentationMixin):
+    """
+    A Provider for using Microsoft Azure Resources
+
+    One of 2 methods are required to authenticate: keyfile, or environment
+    variables. If  keyfile is not set, the following environment
+    variables must be set: `AZURE_CLIENT_ID` (the access key for
+    your azure account),
+    `AZURE_CLIENT_SECRET` (the secret key for your azure account), the
+    `AZURE_TENANT_ID` (the session key for your azure account), and
+    AZURE_SUBSCRIPTION_ID.
+
+    Parameters
+    ----------
+    vm_reference : dict
+        Dictionary describing the parameters of the Azure VM.
+
+        Required structure:
+        {
+          'publisher': VM OS publisher
+          'offer': VM OS offer
+          'sku': VM OS SKU
+          'version': VM OS version
+          'vm_size': VM Size, analogous to instance type for AWS
+          'disk_size_gb': (int) size of VM disk in gb
+          "admin_username": (str) admin username of VM instances,
+          "password": (str) admin password for VM instances
+        }
+    worker_init : str
+        String to append to the Userdata script executed in the cloudinit phase of
+        instance initialization.
+    key_file : str
+        Path to json file that contains 'Azure keys'
+    init_blocks : int
+        Number of blocks to provision at the start of the run. Default is 1.
+    min_blocks : int
+        Minimum number of blocks to maintain. Default is 0.
+    max_blocks : int
+        Maximum number of blocks to maintain. Default is 10.
+    region : str
+        Azure region to launch machines. Default is 'westus'.
+    key_name : str
+        Name of the Azure private key (.pem file) that is usually generated on the console
+        to allow SSH access to the Azure instances. This is mostly used for debugging.
+    launcher : Launcher
+        Launcher for this provider. Possible launchers include
+        :class:`~parsl.launchers.SingleNodeLauncher` (the default),
+        :class:`~parsl.launchers.SrunLauncher`, or
+        :class:`~parsl.launchers.AprunLauncher`
+    linger : Bool
+        When set to True, the workers will not `halt`. The user is responsible for shutting
+        down the nodes.
+    """
+
+    def __init__(self,
+                 init_blocks=1,
+                 min_blocks=0,
+                 max_blocks=10,
+                 parallelism=1,
+                 worker_init='',
+                 vm_reference=None,
```

Comment:
Probably don't assign a default value here unless it actually  makes sense to use that default - I think it never makes senese for `vm_reference` to be None.